### PR TITLE
Stage 5a (4/n): wait out AniList's rate limit instead of failing the sync

### PR DESCRIPTION
**Pull request by HeartlessVeteran2**
## **User description**
Final slice of Stage 5a, following #1232, #1234 and #1235. This was the last of the six planned items that survived verification — the other two dissolved and are written up on #1231.

## What was wrong

AniList allows **90 requests per minute** and answers the 91st with `429 Too Many Requests`. Nothing in the app honoured that.

It is not a hypothetical limit to reach. A library sync walks every tracked manga and issues a `find` plus an `update` for each, so a library of ~45 tracked titles crosses it in a single pass. The first request past the limit failed, `TrackerSyncRepositoryImpl` recorded `SyncStatus.ERROR`, and the user saw a half-finished sync — for a condition the server had explicitly told us how to recover from.

`@TrackerOkHttp` adds a certificate pinner and nothing else; there was no interceptor reading `x-ratelimit-reset` or `Retry-After` anywhere.

## The two headers are not the same units

This is the part worth getting right, because both are "a number of seconds" and they mean opposite things:

| Header | Meaning | Misreading it does what |
|---|---|---|
| `Retry-After` | Seconds **from now** (delta) | Read as a timestamp → a wait in 1970 → immediate re-send |
| `X-RateLimit-Reset` | A Unix **timestamp** (absolute) | Read as a delta → sleep for ~57 years |

Both are converted to a delta and clamped:

- **Ceiling (90s)** — the wait is only as trustworthy as the header it came from. AniList's window is 60 seconds; anything asking meaningfully longer is a misread header or a clock disagreement, and blocking a dispatcher thread on it helps nobody.
- **Floor (1s)** — a clock a few seconds ahead of the server's produces a *negative* delta from `X-RateLimit-Reset`. Without a floor the retry budget would be spent inside a millisecond and change nothing.

A 429 with **no** timing header is returned as-is rather than retried on a guess. A server that didn't say when to come back has given no reason to believe a retry will fare better, and guessing turns one rejected request into several — which is how a client gets rate-limited harder rather than less.

## Where it's registered, and why not the shared client

On the AniList Retrofit instance via `okHttpClient.newBuilder()`, which keeps everything `@TrackerOkHttp` configures — **certificate pinning above all** — and adds one interceptor.

Putting it on `@TrackerOkHttp` itself would apply AniList's 429 semantics to MAL, Kitsu, Shikimori and MangaUpdates, whose limits and headers differ. A retry policy driven by the wrong service's headers is worse than no retry policy.

It's an **application** interceptor rather than a network one: it retries the whole call, and a network interceptor runs per hop, so it wouldn't see the retry as the same logical request.

## Two details that are tested rather than asserted in a comment

Both are the kind of thing that reads as correct and isn't:

- **The 429's body is closed before retrying.** An unclosed body holds its connection out of the pool permanently — invisible at first, compounding with every rate-limited request in a sync, and surfacing much later as connection starvation. I verified the test actually bites by commenting out the `close()` and confirming the connection-count assertion fails (2 of 9 tests red), then restoring it.
- **An interrupt during the wait restores the flag `catch` clears, and fails the call as `IOException`.** Interrupting the thread is how OkHttp cancels a call, so sleeping through it would leave a sync nobody is waiting on holding a dispatcher thread for the rest of the window.

## Verification

`./gradlew :app:assembleDebug testDebugUnitTest detekt ktlintCheck` → **BUILD SUCCESSFUL**.

Nine tests against a real `MockWebServer` rather than a mocked `Chain` — the properties under test are *how many requests actually reach a server and how long apart*, which a mocked chain would assert by construction rather than observe. Covered: a `Retry-After` 429 is retried and the retry's result returned; the wait is at least as long as asked (a test that fails on an implementation retrying instantly); `X-RateLimit-Reset` is treated as a timestamp; a past reset still pauses; a header-less 429 isn't retried; retries are capped at 3 total requests; success passes through; the connection is released; an interrupt fails the call.

## Stage 5a is complete after this

Six planned items → four real, two dissolved:

| Item | Outcome |
|---|---|
| GraphQL variables serialized as strings | Fixed, #1232 |
| `authorizationUrl` missing entirely | Fixed, #1234 (plus the CSRF state that was generated but never sent) |
| `find()` null on missing `mediaListEntry` | Fixed, #1235 |
| `update()` swallowing exceptions | Fixed, #1235 (plus the GraphQL-error paths that returned success) |
| Score normalization | Fixed, #1235 — canonical 0–10 vs AniList's 0–100 |
| Rate limiting | **This PR** |
| `Viewer { id }` for `currentUserId` | Dissolved — field was never read; deleted in #1235 |
| `SaveMediaListEntry` vs `DeleteMediaListEntry` ids | Dissolved — no delete call exists; see #1231 |

Next is **5b**: the metadata layer, Room `manga_metadata` at v41, and `MatchAniListMediaUseCase`.

---
_Generated by [Claude Code](https://claude.ai/code/session_01UoEui2mCCcqArdhyQyFwnR)_

## Summary by Sourcery

Handle AniList rate limiting by retrying 429 responses with bounded waits instead of failing syncs outright.

Enhancements:
- Introduce an AniList-specific OkHttp interceptor that interprets rate-limit headers, waits within safe bounds, and retries 429 responses a limited number of times.
- Wire the AniList Retrofit client to use a derived tracker OkHttpClient with the AniList rate-limit interceptor applied only to AniList traffic.

Tests:
- Add MockWebServer-based tests covering AniList rate-limit handling, including header parsing, minimum/maximum waits, retry capping, connection reuse, success passthrough, and interruption behavior during waits.

___

## **CodeAnt-AI Description**
Retry AniList rate-limit responses instead of failing library syncs

### What Changed
- AniList requests now wait and retry when the service returns a 429 response with a retry time
- Retry timing is read correctly from both relative and absolute server headers, with safe minimum and maximum waits
- Responses without usable timing information remain failed, and retries stop after a limited number of attempts
- Cancelled waits fail promptly without leaving connections occupied; other tracker services keep their existing behavior
- Added coverage for retry timing, retry limits, connection reuse, cancellation, and successful requests

### Impact
`✅ Fewer half-finished AniList library syncs`
`✅ Fewer repeated rate-limit failures`
`✅ No connection leaks during rate-limited syncs`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

---
## Summary by cubic
Retry AniList 429 responses with bounded, cancellable waits to avoid half-finished library syncs. Adds an AniList-only OkHttp interceptor that reads `Retry-After` and `X-RateLimit-Reset` and retries up to 2 times.

- **Bug Fixes**
  - Made waits cancellable: split into 250ms polls with `call().isCanceled()` checks and a monotonic deadline; direct interrupts still fail with `IOException` and restore the interrupt flag.
  - Interprets headers correctly, clamps waits to 1s–90s, closes the 429 body before retrying, and returns 429 when no timing header is present; max 2 retries.
  - Scoped to the AniList Retrofit client via `okHttpClient.newBuilder()`, preserving `@TrackerOkHttp` settings and avoiding cross-service effects; `MockWebServer` tests cover cancellation timing, header parsing, bounds, retry cap, connection reuse, and success passthrough.

<sup>Written for commit ed24e44f7902d4fc4a8dc580ea5236b6146ae954. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/HeartlessVeteran2/Otaku-Reader/pull/1236?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: data/src/main/java/app/otakureader/data/tracking/api/AniListRateLimitInterceptor.kt [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ * does not sit here waiting out a limit nobody is waiting on any more.
+ */
+class AniListRateLimitInterceptor : Interceptor {
+
+    override fun intercept(chain: Interceptor.Chain): Response {
+        var response = chain.proceed(chain.request())
+        var attempt = 0
+
+        while (response.code == HTTP_TOO_MANY_REQUESTS && attempt < MAX_RETRIES) {
+            val waitMs = response.retryDelayMillis() ?: return response
```
[the remark this answers is not in the document]
Comment (HeartlessVeteran2) [reply data/src/main/java/app/otakureader/data/tracking/api/AniListRateLimitInterceptor.kt:62]: Not adding coordination for this one, because the concurrency bound rules the scenario out. Checking rather than assuming, since single-threaded reasoning about concurrent code is a repeat offender in this repo:

- **The caller is sequential.** `TrackerSyncRepositoryImpl.syncAllPending` is a `forEachIndexed` loop that awaits each `syncManga` and additionally sleeps `BATCH_SYNC_STAGGER_MS` between entries. It never fans out.
- **Nothing overrides OkHttp's `Dispatcher`.** `grep -rn "Dispatcher\|maxRequests"` across `core/network` and `data/tracking` returns nothing, so the defaults apply: `maxRequestsPerHost = 5`.

So the ceiling on simultaneous requests to `graphql.anilist.co` is 5. That's the number that would wake together after a shared reset — against a window of **90 requests per minute**. Five requests cannot recreate a 90/minute limit, so there's no herd to disperse and no retry budget to exhaust.

Jitter would be two lines and I considered adding it defensively, but that's mechanism for a case that can't occur at this bound, and it would need a comment claiming to prevent something — which is how a plausible-sounding but false rationale ends up in the file. If the sync loop is ever parallelised, the fix belongs there (a permit-limited fan-out) rather than in per-call retry jitter, since by then the *first* requests would be exhausting the window, not the retries.

Your other finding on this file was correct and is fixed in `ed24e44` — `Call.cancel()` really doesn't interrupt the sleeping thread, my KDoc claimed it did, and the interrupt test passed only because it interrupted the thread itself rather than cancelling the call. The wait is now polled against `chain.call().isCanceled()`, with a test that cancels the `Call` properly and asserts the elapsed time isn't the full 30s. Thanks — that one was a genuine miss.

---
_Generated by [Claude Code](https://claude.ai/code)_
